File: bindings/go/query.go
package powder

import "strings"

// Query is a fluent, injection-safe SELECT builder — the Go mirror of the
// Rust/TypeScript/Java builders. Values go through bound parameters; only
// identifiers you supply are interpolated.
//
//	q := powder.Table("users").Select("id", "name").
//		Filter("score >= ?", 5.0).OrderBy("score", "DESC").Limit(10)
//	batch, err := db.Run(q)
type Query struct {
	table   string
	cols    []string
	wheres  []string
	params  []any
	orderBy string
	limit   *int64
	offset  *int64
}

// Table starts a query against the named table.
func Table(name string) *Query { return &Query{table: name} }

// Select sets the projected columns (defaults to "*").
func (q *Query) Select(columns ...string) *Query {
	q.cols = columns
	return q
}

// Filter adds a WHERE predicate; supply one "?" per parameter. Repeated calls
// are ANDed together.
func (q *Query) Filter(predicate string, params ...any) *Query {
	q.wheres = append(q.wheres, predicate)
	q.params = append(q.params, params...)
	return q
}

// OrderBy sets the sort column and direction ("ASC" or "DESC").
func (q *Query) OrderBy(column, direction string) *Query {
	dir := "ASC"
	if strings.EqualFold(direction, "DESC") {
		dir = "DESC"
	}
	q.orderBy = column + " " + dir
	return q
}

func (q *Query) Limit(n int64) *Query {
	q.limit = &n
	return q
}

func (q *Query) Offset(n int64) *Query {
	q.offset = &n
	return q
}

// SQL renders the statement.
func (q *Query) SQL() string {
	var b strings.Builder
	b.WriteString("SELECT ")
	if len(q.cols) == 0 {
		b.WriteString("*")
	} else {
		b.WriteString(strings.Join(q.cols, ", "))
	}
	b.WriteString(" FROM ")
	b.WriteString(q.table)
	if len(q.wheres) > 0 {
		b.WriteString(" WHERE ")
		b.WriteString(strings.Join(q.wheres, " AND "))
	}
	if q.orderBy != "" {
		b.WriteString(" ORDER BY ")
		b.WriteString(q.orderBy)
	}
	if q.limit != nil {
		b.WriteString(" LIMIT ")
		b.WriteString(itoa(*q.limit))
	}
	if q.offset != nil {
		b.WriteString(" OFFSET ")
		b.WriteString(itoa(*q.offset))
	}
	return b.String()
}

// Params returns the bound parameters, in placeholder order.
func (q *Query) Params() []any { return q.params }

func itoa(n int64) string {
	if n == 0 {
		return "0"
	}
	neg := n < 0
	var buf [20]byte
	i := len(buf)
	u := uint64(n)
	if neg {
		u = uint64(-n)
	}
	for u > 0 {
		i--
		buf[i] = byte('0' + u%10)
		u /= 10
	}
	if neg {
		i--
		buf[i] = '-'
	}
	return string(buf[i:])
}

File: crates/powder-java/java/com/powder/Query.java
package com.powder;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/** Fluent, injection-safe SELECT builder — mirror of the Rust/TS builders. */
public final class Query {
    private final String table;
    private List<String> cols = new ArrayList<>();
    private final List<String> wheres = new ArrayList<>();
    private final List<Object> params = new ArrayList<>();
    private String orderBy;
    private Long limitN;
    private Long offsetN;

    private Query(String table) {
        this.table = table;
    }

    public static Query table(String name) {
        return new Query(name);
    }

    public Query select(String... columns) {
        this.cols = new ArrayList<>(Arrays.asList(columns));
        return this;
    }

    /** Add a WHERE predicate; one {@code ?} per supplied param. ANDed. */
    public Query filter(String predicate, Object... params) {
        this.wheres.add(predicate);
        this.params.addAll(Arrays.asList(params));
        return this;
    }

    public Query order(String column, String direction) {
        this.orderBy = column + " " + ("DESC".equalsIgnoreCase(direction) ? "DESC" : "ASC");
        return this;
    }

    public Query order(String column) {
        return order(column, "ASC");
    }

    public Query limit(long n) {
        this.limitN = n;
        return this;
    }

    public Query offset(long n) {
        this.offsetN = n;
        return this;
    }

    public String sql() {
        StringBuilder sb = new StringBuilder("SELECT ");
        sb.append(cols.isEmpty() ? "*" : String.join(", ", cols));
        sb.append(" FROM ").append(table);
        if (!wheres.isEmpty()) {
            sb.append(" WHERE ").append(String.join(" AND ", wheres));
        }
        if (orderBy != null) {
            sb.append(" ORDER BY ").append(orderBy);
        }
        if (limitN != null) {
            sb.append(" LIMIT ").append(limitN);
        }
        if (offsetN != null) {
            sb.append(" OFFSET ").append(offsetN);
        }
        return sb.toString();
    }

    public Object[] params() {
        return params.toArray();
    }
}
